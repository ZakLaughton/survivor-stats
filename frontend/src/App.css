html,
body,
.root,
.App,
main {
  min-height: 100vh;
}

main {
  text-align: center;
  font-family: 'Londrina Solid', sans-serif;
}

.App-logo {
  animation: App-logo-spin infinite 20s linear;
  height: 40vmin;
}

.App-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.App-link {
  color: #61dafb;
}

article {
  min-height: 100vh;
  display: flex;
  flex-direction: column;
}

article h1 {
  color: white;
  text-shadow: -1px 0 black, 0 2px black, 2px 0 black, 0 -1px black;
  font-family: 'Survivant';
  font-size: 3rem;
  font-weight: 500;
  align-content: center;
  text-align: center;
  margin: 0;
}

@keyframes App-logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}

@font-face {
  font-family: 'Survivant';
  src: url(./fonts/survivant.ttf)
}

@font-face {
  font-family: 'Survivor Font';
  src: url(./fonts/survivorfont.ttf)
}

@import url('https://fonts.googleapis.com/css?family=Londrina+Solid:100,300,400,900');