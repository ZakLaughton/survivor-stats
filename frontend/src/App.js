import React, { Component } from 'react';
import './App.css';
import Tribe from './components/Tribe/Tribe';
import NavBar from './components/NavBar/NavBar';

class App extends Component {
  constructor() {
    super();
    this.state = {
      season: null,
      episode: null,
      castaways: [],
      allTribes: [],
      activeTribes: []
    }
  }

  updateCastaways = async (episode) => {
    // Fetch data
    const url = `http://localhost:3000/?episode=${episode}`;
    const response = await fetch(url);
    const castawayData = await response.json();

    const formattedEpisode = ("0" + episode).slice(-2);


    // Pull unique players
    const allCastaways = castawayData.seasonTribeChanges.map(item => item.castaway);
    const uniqueCastawayStrings = [...new Set(allCastaways)];
    const uniqueCastaways = uniqueCastawayStrings.map(castaway => {return {'name': castaway};})

    // Pull current tribes
    // filter out changes from later episodes
    const currentTribeChanges = castawayData.seasonTribeChanges.filter(change => change.start_episode <= `s37e${formattedEpisode}`)
    // find largest start_episode for each castaway most recent change for each castaway
    const currentCastawaysTribes = uniqueCastaways.map((castaway) => {
      // Find the tribe changes for a given castaway, return the most recent
      const latestCastawayRecord = currentTribeChanges
        .filter(record => {
          return record.castaway === castaway.name;})
        .reduce((prev, current) => {
          return (prev.start_episode > current.start_episode) ? prev : current
      })

      return {
        name: latestCastawayRecord.castaway,
        tribe: latestCastawayRecord.field_value
      }
    })

    const activeTribes = castawayData.tribes.filter((tribe) => {
      if (currentCastawaysTribes.some(castaway => castaway.tribe === tribe.name)) {
        return true;
      } else {
        return false;
      }
      // return tribe/color object
    })

    // Pull unique active tribes
    // const allActiveTribeNames = castawayData.castaways.map(castaway => castaway.tribe);
    // const uniqueActiveTribeNames = [...new Set(allActiveTribeNames)]
    // const activeTribes = castawayData.tribes
    //   .filter((tribe) => uniqueActiveTribeNames.indexOf(tribe.name) > -1);
    
    this.setState({
          castaways: currentCastawaysTribes,
          allTribes: castawayData.tribes,
          activeTribes
          });
  }

  updateEpisode = (event) => {
    const episode = event.target.value;
    this.setState({episode});
    this.updateCastaways(episode);
  }

  async componentDidMount() {
    // await this.updateCastaways()
  }

  render() {
    const {castaways, allTribes, activeTribes, season, episode} = this.state;
    return (
      <div className="App">
        <NavBar 
          season={season}
          episode={episode}
          updateEpisode={this.updateEpisode}
        />
        <main>
          {activeTribes.length > 0 &&
            activeTribes.map(tribe => (
              <Tribe
                key={tribe.name}
                tribe={tribe}
                castaways={castaways} />
            ))
          }
          {activeTribes.length === 0 && 'loading...'}
        </main>
      </div>
    );
  }
}

export default App;
