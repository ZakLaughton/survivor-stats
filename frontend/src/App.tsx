/* eslint-disable jsx-a11y/no-static-element-interactions */
/* eslint-disable jsx-a11y/no-noninteractive-tabindex */
import React, { useEffect, useState, FunctionComponent } from 'react';
import { withRouter, RouteComponentProps } from 'react-router';
import './App.css';
import ReactGA from 'react-ga';
// @ts-ignore Could not find a declaration file for module 'cloudinary-react'.
import { CloudinaryContext } from 'cloudinary-react';
import NavBar from './components/NavBar/NavBar';
// TODO: Set up better prod/dev environment variable strategy
// eslint-disable-next-line no-unused-vars
import { PROD_BACKEND_URL, DEV_BACKEND_URL } from './constants';
// eslint-disable-next-line import/no-unresolved
import { TribeBoard } from './components/TribeBoard/TribeBoard';
import { Episode, ActiveSeasonData, TribalCouncil } from './types';
import SeasonInfoMessage from './components/SeasonInfoMessage/SeasonInfoMessage';
import PreseasonStats from './components/PreseasonStats/PreseasonStats';
import TribalCouncils from './components/TribalCouncils/TribalCouncils';
import { KeyboardEvent } from 'react-native';

const ARRAY_SEARCH_RESULT_NOT_FOUND = -1;
const ARRAY_SORT_KEEP_ORDER = -1;
const ARRAY_SORT_SWAP_ORDER = 1;
const DEFAULT_EPISODE_NUMBER = 0;
const LEFT_ARROW_KEY_CODE = 37;
const RIGHT_ARROW_KEY_CODE = 39;
const SEASON_NUMBER_WITH_PRESEASON_STATS = 38;
const EPISODE_NUMBER_TO_SHOW_PRESEASON_STATS = 0;

interface MatchParams {
  activeSeasonNumber: string;
  activeEpisodeNumber: string;
}

function initializeReactGA() {
  if (document.location.hostname.search(`survivorstats.com`) !== ARRAY_SEARCH_RESULT_NOT_FOUND) {
    ReactGA.initialize(`UA-67511792-3`);
    ReactGA.pageview(`/`);
  }
}

const App: FunctionComponent<RouteComponentProps<MatchParams>> = ({ match, history }) => {
  const activeSeasonNumber = Number(match.params.activeSeasonNumber);
  const activeEpisodeNumber = Number(match.params.activeEpisodeNumber);
  const [activeSeasonData, setActiveSeasonData] = useState<ActiveSeasonData | null>(null);
  const [infoMessage, setInfoMessage] = useState(``);
  useEffect(() => {
    initializeReactGA();
  }, []);

  const isEpisodeActive = (episode: Episode) => episode.active === true;

  useEffect(() => {
    const url = `${PROD_BACKEND_URL}/?season=${activeSeasonNumber}`;
    async function fetchData() {
      const response = await fetch(url);
      setInfoMessage(`Loading...`);
      const newActiveSeasonData: ActiveSeasonData = await response.json();
      newActiveSeasonData.episodes = newActiveSeasonData.episodes
        .filter(isEpisodeActive)
        .sort((a, b) => (a.id > b.id ? ARRAY_SORT_SWAP_ORDER : ARRAY_SORT_KEEP_ORDER));
      setActiveSeasonData(newActiveSeasonData);
      setInfoMessage('');
    }
    fetchData();
  }, [activeSeasonNumber]);

  const currentEpisodeHasTribalCouncils = () => {
    if (activeSeasonData && activeSeasonData.episodes) {
      const episodeData = activeSeasonData.episodes[activeEpisodeNumber];
      if (episodeData && episodeData.tribalCouncils) {
        return !!episodeData.tribalCouncils.length;
      }
    }

    return false;
  };

  const atEarliestEpisode = () => activeEpisodeNumber === DEFAULT_EPISODE_NUMBER;

  const atLatestEpisode = () => {
    if (activeSeasonData && activeSeasonData.episodes) {
      const numberOfEpisodes = activeSeasonData.episodes.length;
      const currentEpisode = Number(activeEpisodeNumber);

      return currentEpisode === numberOfEpisodes - 1;
    }
    return false;
  };

  const decrementEpisode = () => {
    if (!atEarliestEpisode()) {
      const newEpisodeNumber = Number(activeEpisodeNumber) - 1;
      history.push(`/${activeSeasonNumber}/${newEpisodeNumber}`);
    }
  };
  const incrementEpisode = () => {
    if (!atLatestEpisode()) {
      const newEpisodeNumber = Number(activeEpisodeNumber) + 1;
      history.push(`/${activeSeasonNumber}/${newEpisodeNumber}`);
    }
  };

  const onKeyPressed = (e: React.KeyboardEvent) => {
    switch (e.keyCode) {
      case LEFT_ARROW_KEY_CODE:
        decrementEpisode();
        break;
      case RIGHT_ARROW_KEY_CODE:
        incrementEpisode();
        break;
      default:
        break;
    }
  };

  let tribalCouncils: TribalCouncil[] = [];

  if (activeSeasonData && activeSeasonData.episodes) {
    const episodeData = activeSeasonData.episodes[activeEpisodeNumber];
    // eslint-disable-next-line prefer-destructuring
    tribalCouncils = episodeData.tribalCouncils;
  }
  return (
    <CloudinaryContext cloudName='survivorstats'>
      <div className='App' onKeyDown={onKeyPressed} tabIndex={0}>
        <NavBar
          seasonNumber={Number(activeSeasonNumber)}
          episodeNumber={activeEpisodeNumber}
          atEarliestEpisode={atEarliestEpisode}
          atLatestEpisode={atLatestEpisode}
          incrementEpisode={incrementEpisode}
          decrementEpisode={decrementEpisode}
        />
        <div>
          <SeasonInfoMessage message={infoMessage} />
          {
            <main>
              {activeSeasonData && activeSeasonData.episodes && (
                <TribeBoard
                  activeSeasonData={activeSeasonData}
                  activeEpisodeNumber={activeEpisodeNumber}
                />
              )}
              {activeSeasonData &&
                activeSeasonData.preseasonStats &&
                !!activeSeasonData.preseasonStats.length &&
                Number(activeSeasonNumber) === SEASON_NUMBER_WITH_PRESEASON_STATS &&
                Number(activeEpisodeNumber) === EPISODE_NUMBER_TO_SHOW_PRESEASON_STATS && (
                  <PreseasonStats preseasonStats={activeSeasonData.preseasonStats} />
                )}
              {currentEpisodeHasTribalCouncils() && (
                <TribalCouncils seasonNumber={activeSeasonNumber} tribalCouncils={tribalCouncils} />
              )}
            </main>
          }
        </div>
      </div>
    </CloudinaryContext>
  );
};

export default withRouter(App);
