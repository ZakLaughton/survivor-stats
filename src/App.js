import React, { Component } from 'react';
import './App.css';
import Tribe from './components/Tribe/Tribe';

const players = require('./database/players.json');
const tribes = require('./database/tribes.json');
const episodes = require('./database/episodes.json');

class App extends Component {
  render() {
    console.log(players);
    return (
      <div className="App">
        {tribes.map(tribe => (
            <Tribe
              name={tribe.name}
              color={tribe.tribeColor}
              tribeMembers={tribe.members}
              players={players} />
        ))}
      </div>
    );
  }
}

export default App;
