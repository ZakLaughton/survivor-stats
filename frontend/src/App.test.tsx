/* eslint-disable import/no-named-as-default */
/* eslint-disable import/no-named-as-default-member */
/* eslint-env jest */
import React from 'react';
import { shallow } from 'enzyme';
import App from './App';

it('renders without crashing', () => {
  expect(shallow(<App />)).toMatchSnapshot();
});
