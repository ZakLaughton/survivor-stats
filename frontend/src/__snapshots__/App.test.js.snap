// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <NavBar
          allSeasons={Array []}
          atEarliestEpisode={[Function]}
          atLatestEpisode={[Function]}
          decrementEpisode={[Function]}
          episodeId=""
          incrementEpisode={[Function]}
          seasonData={Object {}}
          seasonNum={null}
          setEpisode={[Function]}
          setSeason={[Function]}
        />,
        <Route
          exact={true}
          path="/"
          render={[Function]}
        />,
      ],
      "className": "App",
      "onKeyDown": [Function],
      "tabIndex": "0",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "allSeasons": Array [],
          "atEarliestEpisode": [Function],
          "atLatestEpisode": [Function],
          "decrementEpisode": [Function],
          "episodeId": "",
          "incrementEpisode": [Function],
          "seasonData": Object {},
          "seasonNum": null,
          "setEpisode": [Function],
          "setSeason": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "exact": true,
          "path": "/",
          "render": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <NavBar
            allSeasons={Array []}
            atEarliestEpisode={[Function]}
            atLatestEpisode={[Function]}
            decrementEpisode={[Function]}
            episodeId=""
            incrementEpisode={[Function]}
            seasonData={Object {}}
            seasonNum={null}
            setEpisode={[Function]}
            setSeason={[Function]}
          />,
          <Route
            exact={true}
            path="/"
            render={[Function]}
          />,
        ],
        "className": "App",
        "onKeyDown": [Function],
        "tabIndex": "0",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "allSeasons": Array [],
            "atEarliestEpisode": [Function],
            "atLatestEpisode": [Function],
            "decrementEpisode": [Function],
            "episodeId": "",
            "incrementEpisode": [Function],
            "seasonData": Object {},
            "seasonNum": null,
            "setEpisode": [Function],
            "setSeason": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "exact": true,
            "path": "/",
            "render": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
